### DISCUSSION TITLE
Smarter Receivers for multiple notified resources

### DISCUSSION THREAD

**Discussion by sa-spag** [fluxcd/flux2#2170] [Proposals]
[As brought up on Slack](https://cloud-native.slack.com/archives/CLAJ40HV3/p1638440435270500) (and also partly [here](https://cloud-native.slack.com/archives/CLAJ40HV3/p1638204260149400)), the current Receiver spec and behavior are respectively not very practical and suboptimal when it comes to hooking lots of resources up.

For instance, one may have hundreds of GitHub repositories for micro-services, each containing a single corresponding Helm chart to install through HelmReleases. With the current pattern, one should configure hundreds of webhooks on GitHub and configure hundreds of associated Receivers with their corresponding GitRepositories as `.spec.resources` to notify. This is doable but it seems overly complex to me.

Alternatively, one could provision a single Receiver and hook it up with [a GitHub webhook at the organization level](https://docs.github.com/en/rest/reference/orgs#webhooks). This is simpler to comprehend and setup, but it is far from perfect: the Receiver's `.spec.resources` still needs to be configured by explicitly listing all GitRepositories (as there's no wildcard support), and it leads to most events being irrelevantly handled by downstream controllers (e.g. github.com/org/a-repository sends a request to Receiver/github-org-wide.flux-system which in turns notifies GitRepository/another-repository.flux-system).

To address these issues, I think we could improve the Receiver specification to be more practical without hurting its throughput. Some of the ideas I have in mind:
- make the Receiver smarter so it can filter target resources out according to the received event payload;
- make downstream controllers smarter so they can ignore irrelevant Receivers' notification;
- allow the Receiver's `.spec.resources` to allow `{ kind: "GitRepository", namespace: "*", name: "*" }`, so it's more trivial to set a Receiver up to cover all GitRepositories.

Such solution would be similar to Jenkins' controller webhook: it receives events for all repositories but then routes these to relevant jobs.

**Comment by kingdonb** [comment]
Please take a look at this, it's relevant to your question:

* https://github.com/fluxcd/flux2/pull/2092

Perhaps you have already seen it, but I wanted to link it here since it is directly related to the topic of cross-namespace references.

> hundreds of webhooks on GitHub and configure hundreds of associated Receivers

> Jenkins' controller webhook: it receives events for all repositories but then routes these to relevant jobs.

I have promoted the Jenkins approach before, I second this. It is possible to set up Webhooks once for a given installation, give them all a common target URL (`/github-webhook`), let the administration and the server-side configuration of webhooks happen via GH API (orchestrated internally by Flux) and let a router sort it out based on the content of the webhook POST data and the flux cluster's entire inventory of GitRepos, within some set scope.

It can be problematic for pure multi-tenancy, but surely those issues can be overcome. It might make sense to add a new Receiver type, since this new type we're imagining doesn't really do the same thing as a Receiver, maybe better not to overload the existing class. (Perhaps OrganizationReceiver or something like that.)

**Comment by dm3ch** [comment]
Am I right understand, that Organisation level hooks is still imposible and there's no ongoing work for it now?
Or maybe there's an alternative proposal?

**Reply by kingdonb** [reply]
> there's no ongoing work for it now

Correct, (unless someone is working on this in secret) nobody has brought it up at a Flux dev meeting so far AFAIK.

**Comment by nikmel2803** [comment]
We're actively deploying our services using OCI images via Flux across several k8s clusters. Our typical setup looks like this:

```yaml
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: OCIRepository
metadata:
  name: manifests
spec:
  interval: 24h
  url: oci://ghcr.io/codex-team/test-fluxcd-webhook/manifests
  ref:
    tag: staging
```

The constant need for setting up webhooks for each cluster has been a bit of a pain and prone to errors.

To solve this problem, I wrote a service that listens to webhooks about the publication of new images on GitHub and updates the corresponding sources in all clusters. Thus, it is enough to configure only one webhook for the entire organization so that automatic updating works when a new image is published. For now it supports only OCIRepository sources, but I see no problem to add `GitRepository` as well.

Check out the solution here: [flux-webhook-autoreconciler](https://github.com/codex-team/flux-webhook-autoreconciler). We've already tested it in a real-world setup and it works well. I'd love to hear what you think or see how it works for you if you try it out!
